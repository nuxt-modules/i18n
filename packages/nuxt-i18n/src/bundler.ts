import createDebug from 'debug'
import { resolve } from 'pathe'
import { vueI18n } from '@intlify/vite-plugin-vue-i18n'
import { extendViteConfig, extendWebpackConfig } from '@nuxt/kit'

type VitePluginOptions = Parameters<typeof vueI18n>[0]

const debug = createDebug('@nuxtjs/i18n:bundler')

export async function extendBundler(hasLocaleFiles: boolean, langPath: string | null) {
  try {
    // @ts-ignore NOTE: use webpack which is installed by nuxt
    const webpack = await import('webpack').then(m => m.default || m)

    // TODO: vue-i18n-loader cannot be resolved as i18n resources which compiled by the intlify message compiler ...
    extendWebpackConfig(config => {
      if (hasLocaleFiles && langPath) {
        config.module!.rules!.push({
          test: /\.(json5?|ya?ml)$/,
          type: 'javascript/auto',
          loader: '@intlify/vue-i18n-loader',
          include: [resolve(langPath, './**')]
        })
      }

      config.module!.rules!.push({
        resourceQuery: /blockType=i18n/,
        type: 'javascript/auto',
        loader: '@intlify/vue-i18n-loader'
      })

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- `config.plugins` is safe, so it's assigned with nuxt!
      config.plugins!.push(
        new webpack.DefinePlugin({
          __VUE_I18N_FULL_INSTALL__: 'true',
          __VUE_I18N_LEGACY_API__: 'true',
          __INTLIFY_PROD_DEVTOOLS__: 'false'
        })
      )
    })
  } catch (e: unknown) {
    debug((e as Error).message)
  }

  // install @intlify/vite-plugin-vue-i18n
  extendViteConfig(config => {
    const viteOptions: VitePluginOptions = {
      compositionOnly: false
    }
    if (hasLocaleFiles && langPath) {
      viteOptions['include'] = resolve(langPath, './**')
    }
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- `config.plugins` is safe, so it's assigned with nuxt!
    config.plugins!.push(vueI18n(viteOptions))
  })
}
