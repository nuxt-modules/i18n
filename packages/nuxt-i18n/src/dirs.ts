import { fileURLToPath } from 'url'
import { dirname, resolve } from 'pathe'

export const distDir = dirname(fileURLToPath(import.meta.url))
const pkgDir = resolve(distDir, '..')
export const pkgModulesDir = resolve(pkgDir, './node_modules')
