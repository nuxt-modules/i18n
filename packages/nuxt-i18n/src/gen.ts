import createDebug from 'debug'
import { isString, isRegExp, isFunction, isArray, isObject } from '@intlify/shared'
import { templateUtils } from '@nuxt/kit'
import { genImport } from 'knitwork'

import type { NuxtI18nOptions, NuxtI18nInternalOptions, LocaleInfo, NoNullable } from './types'
import type { NuxtI18nOptionsDefault } from './constants'

export type LoaderOptions = {
  localeCodes?: string[]
  localeInfo?: LocaleInfo[]
  nuxtI18nOptions?: NuxtI18nOptions
  nuxtI18nOptionsDefault?: NuxtI18nOptionsDefault
  nuxtI18nInternalOptions?: NuxtI18nInternalOptions
}

const debug = createDebug('@nuxtjs/i18n:gen')

export function generateLoaderOptions(
  lazy: NoNullable<NuxtI18nOptions['lazy']>,
  langDir: NuxtI18nOptions['langDir'],
  options: LoaderOptions = {}
) {
  // TODO: lazy loading local info
  let genCode = ''
  const localeInfo = options.localeInfo || []
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const { fallbackLocale } = (isObject(options.nuxtI18nOptions?.vueI18n) ? options.nuxtI18nOptions?.vueI18n : {}) as any
  const syncLocaleFiles = new Set<LocaleInfo>()
  const asyncLocaleFiles = new Set<LocaleInfo>()

  if (langDir) {
    if (fallbackLocale && isString(fallbackLocale)) {
      const localeObject = localeInfo.find(l => l.code === fallbackLocale)
      if (localeObject) {
        syncLocaleFiles.add(localeObject)
      }
    }
    for (const locale of localeInfo) {
      if (!syncLocaleFiles.has(locale) && !asyncLocaleFiles.has(locale)) {
        ;(lazy ? asyncLocaleFiles : syncLocaleFiles).add(locale)
      }
    }
  }

  const importMapper = new Map<string, string>()
  for (const { code, path } of syncLocaleFiles) {
    importMapper.set(code, templateUtils.importName(`locale_${code}`))
    genCode += `${genImport(path, templateUtils.importName(`locale_${code}`))}\n`
  }

  // prettier-ignore
  genCode += `${Object.entries(options).map(([rootKey, rootValue]) => {
    if (rootKey === 'nuxtI18nOptions') {
      let genCodes = `export const resolveNuxtI18nOptions = async (context) => {\n`
      genCodes += `  const ${rootKey} = Object({})\n`
      for (const [key, value] of Object.entries(rootValue)) {
        if (key === 'vueI18n') {
          const optionLoaderVariable = `${key}OptionsLoader`
          genCodes += `  const ${optionLoaderVariable} = ${isObject(value)
            ? `async (context) => ${toCode(value)}\n`
            : isString(value)
              ? `async (context) => import(${toCode(value)}).then(r => (r.default || r)(context))\n`
              : `async (context) => ${toCode({})}\n`
          }`
          genCodes += `  ${rootKey}.${key} = await ${optionLoaderVariable}(context)\n`
        } else {
          genCodes += `  ${rootKey}.${key} = ${toCode(value)}\n`
        }
      }
      genCodes += `  return nuxtI18nOptions\n`
      genCodes += `}\n`
      return genCodes
    } else if (rootKey === 'nuxtI18nOptionsDefault') {
      // generate default nuxtI18n options
      return `export const ${rootKey} = Object({${Object.entries(rootValue).map(([key, value]) => {
        return `${key}: ${toCode(value)}`
      }).join(`,`)}})\n`
    } else if (rootKey === 'nuxtI18nInternalOptions') {
      return `export const ${rootKey} = Object({${Object.entries(rootValue).map(([key, value]) => {
        return `${key}: ${toCode(value)}`
      }).join(`,`)}})\n`
    } else if (rootKey === 'localeInfo') {
      let codes = `export const loadMessages = async () => {\n`
      codes += `  const messages = Object({})\n`
      if (langDir) {
        for (const { code } of syncLocaleFiles) {
          codes += `  messages[${toCode(code)}] = ${importMapper.get(code)}\n`
        }
        for (const { code, path } of asyncLocaleFiles) {
          codes += `  messages[${toCode(code)}] = await import(${toCode(path)} /* webpackChunkName: ${toCode(path)} */).then(r => r.default || r)\n`
        }
      }
      codes += `  return Promise.resolve(messages)\n`
      codes += `}\n`
      return codes
	  } else {
	    return `export const ${rootKey} = ${toCode(rootValue)}\n`
	  }
  }).join('\n')}`

  debug('generate code', genCode)
  return genCode
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function stringifyObj(obj: Record<string, any>): string {
  return `Object({${Object.entries(obj)
    .map(([key, value]) => `${JSON.stringify(key)}:${toCode(value)}`)
    .join(`,`)}})`
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function toCode(code: any): string {
  if (code === null) {
    return `null`
  }

  if (code === undefined) {
    return `undefined`
  }

  if (isString(code)) {
    return JSON.stringify(code)
  }

  if (isRegExp(code) && code.toString) {
    return code.toString()
  }

  if (isFunction(code) && code.toString) {
    return `(${code.toString()})`
  }

  if (isArray(code)) {
    return `[${code.map(c => toCode(c)).join(`,`)}]`
  }

  if (isObject(code)) {
    return stringifyObj(code)
  }

  return code + ``
}
