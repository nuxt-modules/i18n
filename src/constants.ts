export const NUXT_I18N_MODULE_ID = '@nuxtjs/i18n'
export const VUE_I18N_PKG = 'vue-i18n'
export const SHARED_PKG = '@intlify/shared'
export const MESSAGE_COMPILER_PKG = '@intlify/message-compiler'
export const CORE_PKG = '@intlify/core'
export const CORE_BASE_PKG = '@intlify/core-base'
export const H3_PKG = '@intlify/h3'
export const UTILS_PKG = '@intlify/utils'
export const UTILS_H3_PKG = '@intlify/utils/h3'
export const UFO_PKG = 'ufo'
export const IS_HTTPS_PKG = 'is-https'

// Options
const STRATEGY_PREFIX = 'prefix'
const STRATEGY_PREFIX_EXCEPT_DEFAULT = 'prefix_except_default'
const STRATEGY_PREFIX_AND_DEFAULT = 'prefix_and_default'
const STRATEGY_NO_PREFIX = 'no_prefix'
export const STRATEGIES = {
  PREFIX: STRATEGY_PREFIX,
  PREFIX_EXCEPT_DEFAULT: STRATEGY_PREFIX_EXCEPT_DEFAULT,
  PREFIX_AND_DEFAULT: STRATEGY_PREFIX_AND_DEFAULT,
  NO_PREFIX: STRATEGY_NO_PREFIX
} as const

export const DEFAULT_DYNAMIC_PARAMS_KEY = 'nuxtI18n'
export const DEFAULT_COOKIE_KEY = 'i18n_redirected'
export const SWITCH_LOCALE_PATH_LINK_IDENTIFIER = 'nuxt-i18n-slp'

export const DEFAULT_OPTIONS = {
  experimental: {
    localeDetector: '',
    switchLocalePathLinkSSR: false,
    autoImportTranslationFunctions: false
  },
  bundle: {
    compositionOnly: true,
    runtimeOnly: false,
    fullInstall: true,
    dropMessageCompiler: false
  },
  compilation: {
    jit: true,
    strictMessage: true,
    escapeHtml: false
  },
  customBlocks: {
    defaultSFCLang: 'json',
    globalSFCScope: false
  },
  vueI18n: '',
  locales: [] as string[],
  defaultLocale: '',
  defaultDirection: 'ltr',
  routesNameSeparator: '___',
  trailingSlash: false,
  defaultLocaleRouteNameSuffix: 'default',
  strategy: STRATEGY_PREFIX_EXCEPT_DEFAULT,
  lazy: false,
  langDir: null,
  rootRedirect: undefined,
  detectBrowserLanguage: {
    alwaysRedirect: false,
    cookieCrossOrigin: false,
    cookieDomain: null,
    cookieKey: DEFAULT_COOKIE_KEY,
    cookieSecure: false,
    fallbackLocale: '',
    redirectOn: 'root',
    useCookie: true
  },
  differentDomains: false,
  baseUrl: '',
  dynamicRouteParams: false,
  customRoutes: 'page',
  pages: {},
  skipSettingLocaleOnNavigate: false,
  types: 'composition',
  debug: false,
  parallelPlugin: false
} as const

export const NUXT_I18N_TEMPLATE_OPTIONS_KEY = 'i18n.options.mjs'
export const NUXT_I18N_COMPOSABLE_DEFINE_ROUTE = 'defineI18nRoute'
export const NUXT_I18N_COMPOSABLE_DEFINE_LOCALE = 'defineI18nLocale'
export const NUXT_I18N_COMPOSABLE_DEFINE_CONFIG = 'defineI18nConfig'
export const NUXT_I18N_COMPOSABLE_DEFINE_LOCALE_DETECTOR = 'defineI18nLocaleDetector'

export const TS_EXTENSIONS = ['.ts', '.cts', '.mts']
export const JS_EXTENSIONS = ['.js', '.cjs', '.mjs']
export const EXECUTABLE_EXTENSIONS = [...JS_EXTENSIONS, ...TS_EXTENSIONS]

export const NULL_HASH = '00000000'
