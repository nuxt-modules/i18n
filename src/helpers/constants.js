/**
 * @typedef {import('../../types').Options} Options
 */

// Internals
export const ROOT_DIR = 'nuxt-i18n'

// Options
const STRATEGY_PREFIX = 'prefix'
const STRATEGY_PREFIX_EXCEPT_DEFAULT = 'prefix_except_default'
const STRATEGY_PREFIX_AND_DEFAULT = 'prefix_and_default'
const STRATEGY_NO_PREFIX = 'no_prefix'
export const STRATEGIES = {
  PREFIX: STRATEGY_PREFIX,
  PREFIX_EXCEPT_DEFAULT: STRATEGY_PREFIX_EXCEPT_DEFAULT,
  PREFIX_AND_DEFAULT: STRATEGY_PREFIX_AND_DEFAULT,
  NO_PREFIX: STRATEGY_NO_PREFIX
}

export const COMPONENT_OPTIONS_KEY = 'nuxtI18n'

/** @type {Options} */
export const DEFAULT_OPTIONS = {
  vueI18n: {},
  vueI18nLoader: false,
  locales: [],
  defaultLocale: '',
  defaultDirection: 'ltr',
  routesNameSeparator: '___',
  defaultLocaleRouteNameSuffix: 'default',
  sortRoutes: true,
  strategy: STRATEGY_PREFIX_EXCEPT_DEFAULT,
  lazy: false,
  langDir: null,
  rootRedirect: null,
  detectBrowserLanguage: {
    alwaysRedirect: false,
    cookieCrossOrigin: false,
    cookieDomain: null,
    cookieKey: 'i18n_redirected',
    cookieSecure: false,
    fallbackLocale: '',
    onlyOnNoPrefix: false,
    onlyOnRoot: false,
    useCookie: true
  },
  differentDomains: false,
  seo: false,
  baseUrl: '',
  vuex: {
    moduleName: 'i18n',
    syncLocale: false,
    syncMessages: false,
    syncRouteParams: true
  },
  parsePages: true,
  pages: {},
  skipSettingLocaleOnNavigate: false,
  beforeLanguageSwitch: () => null,
  onBeforeLanguageSwitch: () => {},
  onLanguageSwitched: () => null
}
