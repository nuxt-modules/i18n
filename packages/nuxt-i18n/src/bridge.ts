import { resolve } from 'pathe'
import { resolveModule, addPluginTemplate } from '@nuxt/kit'
import { distDir } from './dirs'

import type { Nuxt } from '@nuxt/schema'

export async function setupNuxtBridge(nuxt: Nuxt) {
  // resolve vue-i18n as vue-i18n-legacy
  nuxt.options.alias['vue-i18n'] = resolveModule('vue-i18n-legacy/dist/vue-i18n.esm.js', {
    paths: nuxt.options.modulesDir
  })
  nuxt.options.build.transpile.push('vue-i18n')

  // resolve vue-i18n-bridge as vue-i18n
  const vueI18nBridgePath = nuxt.options.dev
    ? 'vue-i18n-bridge/dist/vue-i18n-bridge.esm-bundler.js'
    : 'vue-i18n-bridge/dist/vue-i18n-bridge.runtime.esm-bundler.js'
  nuxt.options.alias['vue-i18n-bridge'] = resolveModule(vueI18nBridgePath, {
    paths: nuxt.options.modulesDir
  })
  nuxt.options.build.transpile.push('vue-i18n-bridge')

  addPluginTemplate({
    filename: 'runtime/bridge.plugin.mjs',
    src: resolve(distDir, 'runtime/bridge.plugin.mjs')
  })
}
