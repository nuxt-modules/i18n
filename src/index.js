import { resolve, join } from 'path'
import { readdirSync } from 'fs'
import merge from 'lodash.merge'
// @ts-ignore
import { directive as i18nExtensionsDirective } from '@intlify/vue-i18n-extensions'
import { COMPONENT_OPTIONS_KEY, DEFAULT_OPTIONS, ROOT_DIR, STRATEGIES, REDIRECT_ON_OPTIONS } from './helpers/constants'
import { buildHook, createExtendRoutesHook } from './core/hooks'
import { formatMessage } from './templates/utils-common'

/** @type {import('@nuxt/types').Module<import('../types').Options>} */
export default async function (moduleOptions) {
  /** @type {import('../types/internal').ResolvedOptions} */
  const options = merge({}, DEFAULT_OPTIONS, moduleOptions, this.options.i18n)

  if (!Object.values(STRATEGIES).includes(options.strategy)) {
    // eslint-disable-next-line no-console
    console.error(formatMessage(`Invalid "strategy" option "${options.strategy}" (must be one of: ${Object.values(STRATEGIES).join(', ')}).`))
    return
  }

  if (options.lazy && !options.langDir) {
    throw new Error(formatMessage('When using the "lazy" option you must also set the "langDir" option.'))
  }

  if (options.langDir) {
    if (!options.locales.length || typeof options.locales[0] === 'string') {
      throw new Error(formatMessage('When using the "langDir" option the "locales" must be a list of objects.'))
    }
    for (const locale of options.locales) {
      if (typeof (locale) === 'string' || !locale.file) {
        throw new Error(formatMessage(`All locales must be objects and have the "file" property set when using "langDir".\nFound none in:\n${JSON.stringify(locale, null, 2)}.`))
      }
    }
    options.langDir = this.nuxt.resolver.resolveAlias(options.langDir)
  }

  // Templates (including plugins).
  // This is done here rather than in the build hook to ensure the order the plugins are added
  // is predictable between different modules.
  const nuxtOptions = this.options
  const normalizedLocales = []
  for (const locale of options.locales) {
    if (typeof (locale) === 'string') {
      normalizedLocales.push({ code: locale })
    } else {
      normalizedLocales.push(locale)
    }
  }
  options.normalizedLocales = normalizedLocales
  // Get an array of locale codes from the list of locales.
  options.localeCodes = options.normalizedLocales.map(locale => locale.code)

  const templatesOptions = {
    Constants: {
      COMPONENT_OPTIONS_KEY,
      STRATEGIES,
      REDIRECT_ON_OPTIONS
    },
    nuxtOptions: {
      isUniversalMode: nuxtOptions.mode === 'universal',
      trailingSlash: nuxtOptions.router.trailingSlash
    },
    options
  }

  const templatesPath = join(__dirname, '/templates')
  for (const file of readdirSync(templatesPath)) {
    if (!file.endsWith('.js')) {
      continue
    }
    if (file.startsWith('plugin.')) {
      this.addPlugin({
        src: resolve(templatesPath, file),
        fileName: join(ROOT_DIR, file),
        options: templatesOptions
      })
    } else {
      this.addTemplate({
        src: resolve(templatesPath, file),
        fileName: join(ROOT_DIR, file),
        options: templatesOptions
      })
    }
  }

  if (options.strategy !== STRATEGIES.NO_PREFIX && options.localeCodes.length) {
    this.extendRoutes(createExtendRoutesHook.call(this, options))
  }

  await this.nuxt.hook('build:before', () => buildHook.call(this, options))

  this.options.alias['~i18n-klona'] = require.resolve('klona/full').replace(/\.js$/, '.mjs')
  this.options.alias['~i18n-ufo'] = require.resolve('ufo').replace(/\.js$/, '.mjs')

  if (!Array.isArray(this.options.router.middleware)) {
    throw new TypeError(formatMessage('options.router.middleware is not an array.'))
  }
  this.options.router.middleware.push('nuxti18n')

  if (!this.options.render.bundleRenderer || typeof (this.options.render.bundleRenderer) !== 'object') {
    throw new TypeError(formatMessage('options.render.bundleRenderer is not an object.'))
  }
  this.options.render.bundleRenderer.directives = this.options.render.bundleRenderer.directives || {}
  this.options.render.bundleRenderer.directives.t = i18nExtensionsDirective

  // Transpile is-https (IE11)
  this.options.build.transpile = this.options.build.transpile || []
  this.options.build.transpile.push('is-https')
}
