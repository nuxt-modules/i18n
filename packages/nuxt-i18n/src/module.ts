import createDebug from 'debug'
import { isObject, isString } from '@intlify/shared'
import { defineNuxtModule, isNuxt2, isNuxt3, getNuxtVersion, resolveModule, addTemplate } from '@nuxt/kit'
import { resolve } from 'pathe'
import defu from 'defu'
import { setupNuxtBridge } from './bridge'
import { setupAutoImports } from './auto-imports'
import { extendBundler } from './bundler'
import { setupNuxt3 } from './nuxt3'
import { DEFAULT_OPTIONS, STRATEGIES } from './constants'
import { getNormalizedLocales, resolveLocales } from './utils'
import { setupPages } from './pages'
import { generateLoaderOptions } from './gen'

import type { Composer } from 'vue-i18n'
import type { NuxtI18nOptions } from './types'

export * from './types'

const debug = createDebug('@nuxt/i18n:module')

export default defineNuxtModule<NuxtI18nOptions>({
  meta: {
    name: '@nuxt/i18n',
    configKey: 'i18n'
  },
  defaults: {},
  async setup(i18nOptions, nuxt) {
    const options = defu(i18nOptions, DEFAULT_OPTIONS) as Required<NuxtI18nOptions>
    options.langDir = options.langDir || 'locales'
    debug('options', options)

    if (isNuxt2(nuxt)) {
      await setupNuxtBridge(nuxt)
    } else if (isNuxt3(nuxt)) {
      await setupNuxt3(nuxt)
    } else {
      throw new Error(`Cannot support nuxt version: ${getNuxtVersion(nuxt)}`)
    }

    // resolve langDir
    const langPath = resolve(nuxt.options.srcDir, options.langDir)
    debug('langDir path', langPath)

    // resolve localeInfo
    const normalizedLocales = (options.__normalizedLocales = getNormalizedLocales(options.locales))
    const hasLocaleFiles = normalizedLocales.length > 0
    const localeCodes = normalizedLocales.map(locale => locale.code)
    const localeInfo = await resolveLocales(langPath, normalizedLocales)
    debug('localeInfo', localeInfo)

    // resolve vueI18n options
    // prettier-ignore
    options.vueI18n = isObject(options.vueI18n)
      ? options.vueI18n
      : isString(options.vueI18n)
        ? resolve(nuxt.options.rootDir, options.vueI18n)
        : {}

    // setup nuxt/pages
    if (options.strategy !== STRATEGIES.NO_PREFIX && localeCodes.length) {
      await setupPages(options, nuxt, { isBridge: isNuxt2(nuxt), localeCodes })
    }

    // resolve @intlify/shared
    nuxt.options.alias['@intlify/shared'] = resolveModule('@intlify/shared/dist/shared.esm-bundler.js', {
      paths: nuxt.options.modulesDir
    })
    nuxt.options.build.transpile.push('@intlify/shared')

    // resolve vue-i18n-routing
    nuxt.options.alias['vue-i18n-routing'] = resolveModule('vue-i18n-routing/dist/vue-i18n-routing.es.js', {
      paths: nuxt.options.modulesDir
    })
    nuxt.options.build.transpile.push('vue-i18n-routing')

    // loading options template
    addTemplate({
      filename: 'i18n.options.mjs',
      getContents: () => {
        return generateLoaderOptions({ localeCodes, localeInfo, nuxtI18nOptions: options })
      }
    })

    // setup auto imports
    await setupAutoImports(nuxt)

    // extend bundler
    await extendBundler(hasLocaleFiles, langPath)
  }
})

declare module '@nuxt/kit' {
  export interface NuxtApp {
    $i18n: Composer
  }
}

declare module '@nuxt/schema' {
  interface NuxtConfig {
    i18n?: NuxtI18nOptions
  }
}
