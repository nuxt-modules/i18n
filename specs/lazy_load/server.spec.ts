import { test, expect, describe } from 'vitest'
import { fileURLToPath } from 'node:url'
import { setup, $fetch, fetch } from '@nuxt/test-utils'
import { validateSyntax } from '../helper'
import { NUXT_I18N_PRECOMPILE_ENDPOINT } from '../../src/constants'

await setup({
  rootDir: fileURLToPath(new URL(`../fixtures/lazy`, import.meta.url)),
  browser: true,
  // overrides
  nuxtConfig: {
    i18n: {
      experimental: {
        jsTsFormatResource: true
      },
      defaultLocale: 'en',
      langDir: 'lang',
      lazy: true,
      locales: [
        {
          code: 'en',
          iso: 'en-US',
          file: 'en.json',
          name: 'English'
        },
        {
          code: 'en-GB',
          iso: 'en-GB',
          files: ['en.json', 'en-GB.js', 'en-GB.ts'],
          name: 'English (UK)'
        },
        {
          code: 'fr',
          iso: 'fr-FR',
          file: 'fr.json5',
          name: 'Français'
        }
      ]
    }
  }
})

describe('success', async () => {
  test('locale', async () => {
    const code = await $fetch(NUXT_I18N_PRECOMPILE_ENDPOINT, {
      method: 'POST',
      body: {
        type: 'locale',
        locale: 'en',
        resource: {
          hello: 'Hello'
        }
      }
    })
    expect(validateSyntax(code)).toBe(true)
    expect(code).toMatchSnapshot()
  })

  test('config', async () => {
    const code = await $fetch(NUXT_I18N_PRECOMPILE_ENDPOINT, {
      method: 'POST',
      body: {
        type: 'config',
        configId: '1234',
        resource: {
          ja: {
            hello: 'こんにちは'
          }
        }
      }
    })
    expect(validateSyntax(code)).toBe(true)
    expect(code).toMatchSnapshot()
  })
})

describe('fail', () => {
  test('no type param', async () => {
    const res = await fetch(NUXT_I18N_PRECOMPILE_ENDPOINT, {
      method: 'POST',
      body: JSON.stringify({
        resource: {
          hello: 'Hello'
        }
      })
    })
    expect(res.status).toBe(400)
    expect(res.statusText).toBe(`require the 'type'`)
  })

  test('no configId param', async () => {
    const res = await fetch(NUXT_I18N_PRECOMPILE_ENDPOINT, {
      method: 'POST',
      body: JSON.stringify({
        type: 'config',
        resource: {
          hello: 'Hello'
        }
      })
    })
    expect(res.status).toBe(400)
    expect(res.statusText).toBe(`require the 'configId'`)
  })

  test('no locale param', async () => {
    const res = await fetch(NUXT_I18N_PRECOMPILE_ENDPOINT, {
      method: 'POST',
      body: JSON.stringify({
        type: 'locale',
        resource: {
          hello: 'Hello'
        }
      })
    })
    expect(res.status).toBe(400)
    expect(res.statusText).toBe(`require the 'locale'`)
  })

  test('no resource param', async () => {
    const res = await fetch(NUXT_I18N_PRECOMPILE_ENDPOINT, {
      method: 'POST',
      body: JSON.stringify({
        type: 'locale',
        locale: 'en'
      })
    })
    expect(res.status).toBe(400)
    expect(res.statusText).toBe(`require the 'resource'`)
  })

  test('include html code in resource', async () => {
    const res = await fetch(NUXT_I18N_PRECOMPILE_ENDPOINT, {
      method: 'POST',
      body: JSON.stringify({
        type: 'locale',
        locale: 'en-tag',
        resource: {
          hello: '<script>window.alert("seciruty issue")</script>'
        }
      })
    })
    expect(res.status).toBe(400)
    expect(res.statusText).toBe(`Detected HTML in '<script>window.alert("seciruty issue")</script>' message.`)
  })
})
