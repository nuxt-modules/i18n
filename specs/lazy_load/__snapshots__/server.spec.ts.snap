// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`success > config 1`] = `
"export default {
  \\"ja\\": {
  \\"hello\\": (ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"こんにちは\\"])}
},
}
"
`;

exports[`success > locale 1`] = `
"const resource = {
  \\"hello\\": (ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"Hello\\"])}
}
export default resource"
`;
